onadata/apps/api/viewsets/tableau_viewset.py

```diff
@@ -0,0 +1,288 @@
+from rest_framework.decorators import action
+from rest_framework.response import Response
+
+from onadata.libs.data import parse_int
+from onadata.libs.renderers.renderers import pairing
+from onadata.apps.logger.models import Instance
+from onadata.apps.logger.models.xform import XForm
+from onadata.apps.viewer.models.data_dictionary import DataDictionary
+from onadata.apps.api.viewsets.open_data_viewset import (
+    OpenDataViewSet)
+from onadata.libs.serializers.data_serializer import TableauDataSerializer
+from onadata.libs.utils.common_tags import (
+    ID, MULTIPLE_SELECT_TYPE, REPEAT_SELECT_TYPE, PARENT_TABLE, PARENT_ID)
+
+
+DEFAULT_TABLE_NAME = 'data'
+gps_parts = ['altitude', 'precision', 'longitude', 'latitude']
+
+
+def process_tableau_data(
+        data, xform,
+        parent_table: str = None,
+        parent_id: int = None,
+        current_table: str = DEFAULT_TABLE_NAME):
+    result = []
+    if data:
+        for idx, row in enumerate(data, start=1):
+            flat_dict = defaultdict(dict)
+            row_id = row.get('_id')
+
+            if not row_id and parent_id:
+                row_id = int(pairing(parent_id, idx))
+                flat_dict[PARENT_ID] = parent_id
+                flat_dict[PARENT_TABLE] = parent_table
+                flat_dict[ID] = row_id
+            else:
+                flat_dict[ID] = row_id
+
+            for (key, value) in row.items():
+                qstn = xform.get_element(key)
+                if qstn:
+                    qstn_type = qstn.get('type')
+                    qstn_name = qstn.get('name')
+
+                    prefix_parts = [
+                        question['name'] for question in qstn.get_lineage()
+                        if question['type'] == 'group'
+                    ]
+                    prefix = "_".join(prefix_parts)
+
+                    if qstn_type == REPEAT_SELECT_TYPE:
+                        repeat_data = process_tableau_data(
+                            value, xform,
+                            parent_table=current_table,
+                            parent_id=row_id,
+                            current_table=qstn_name)
+                        # Pop any list within the returned repeat data.
+                        # Lists represent a repeat group which should be in a
+                        # separate field.
+                        cleaned_data = []
+                        for data_dict in repeat_data:
+                            remove_keys = []
+                            for k, v in data_dict.items():
+                                if isinstance(v, list):
+                                    remove_keys.append(k)
+                                    if not flat_dict[k]:
+                                        flat_dict[k] = v
+                                    else:
+                                        flat_dict[k].extend(v)
+                            [data_dict.pop(k) for k in remove_keys]
```

Comment:
Codacy found an issue: [Expression "[data_dict.pop(k) for k in remove_keys]" is assigned to nothing](https://app.codacy.com/gh/onaio/onadata/pullRequest?prid=6260785)